From 366013860c3c04e4a37216857d93495048ff8bc3 Mon Sep 17 00:00:00 2001
From: Jan Sondhauss <jan.sondhauss@wago.com>
Date: Thu, 16 Apr 2020 12:04:56 +0200
Subject: [PATCH] net: phy:_phy_device: set phydev state properly

Upon phy attachment the pyhdev state has to reflect the actual state of the phy.
When the phy is powered up (i.e. resumed) the phydev should be in READY state,
when the phy is powered down (i.e. suspended) the phydev should be in HALTED state.

Signed-off-by: Jan Sondhauss <jan.sondhauss@wago.com>
---
 drivers/net/phy/phy_device.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/drivers/net/phy/phy_device.c b/drivers/net/phy/phy_device.c
index fb9e1c8..a3668a2 100644
--- a/drivers/net/phy/phy_device.c
+++ b/drivers/net/phy/phy_device.c
@@ -917,7 +917,6 @@ int phy_attach_direct(struct net_device *dev, struct phy_device *phydev,
 
 	phydev->interface = interface;
 
-	phydev->state = PHY_READY;
 
 	/* Initial carrier state is off as the phy is about to be
 	 * (re)initialized.
@@ -929,12 +928,15 @@ int phy_attach_direct(struct net_device *dev, struct phy_device *phydev,
 	 * (dev_flags and interface)
 	 */
 	err = phy_init_hw(phydev);
-	if (err)
+	if (err) {
 		phy_detach(phydev);
-	else if (dev->flags & IFF_UP)
+	} else if (dev->flags & IFF_UP) {
+		phydev->state = PHY_READY;
 		phy_resume(phydev);
-	else
+	} else {
+		phydev->state = PHY_HALTED;
 		phy_suspend(phydev);
+	}
 
 	return err;
 
-- 
2.7.4

